Introduce `deer` crate at v0.0.0 (#1163)
hashintel/hash: packages/libs/deer/README.md

> # deer

Reviewer: Purposefully lower case?

Author: Yes, to keep in line with error-stack and how Rust crate naming usually is.

Reviewer: I appreciate that the crate names are usually lower-case, but when referring to them in headings I haven't seen them left as lower-case before.

See:
- [Tokio](https://github.com/tokio-rs/tokio)
- Tokio [Tracing](https://github.com/tokio-rs/tracing)
- [Serde](https://github.com/serde-rs/serde)
- [Serde JSON](https://github.com/serde-rs/json)

`deer` is a backend agnostic deserialization framework, featuring meaningful error messages and context (utilizing [`error-stack`](https://crates.io/crates/error-stack)) and a fail-slow behavior by default.

## Fail-Slow

> Currently available Rust deserializers have mostly been developed with correctness and speed in mind. These are universally beneficial optimizations, but in certain cases (such as when collecting user-facing validation feedback) there are relatively few options available within Rust that allow for extended evaluation beyond a single error. `deer`aims to improve this situation by consciously trading off an acceptable degree of speed to enable the surfacing of multiple errors.

Reviewer: ```suggestion
Currently available Rust deserializers have mostly been developed with correctness and speed in mind. These are universally beneficial optimizations, but in certain cases (such as when collecting user-facing validation feedback) there are relatively few options available within Rust that allow for extended evaluation beyond a single error. `deer` aims to improve this situation by consciously trading off an acceptable degree of speed to enable the surfacing of multiple errors.
```

## Example

> A popular example for `deer` are end-user facing APIs, given the following example:

Reviewer: ```suggestion
End-user facing APIs are a well-suited example for `deer`. 

Given the following example:
```

```rust
#[derive(Debug, serde::Deserialize, deer::Deserialize)]
struct Body {
    u8: u8,
    string: String
}

fn main() {
    let payload = json!({
        "u8": 256,
        "string": null,
        "extra": 1
    });

    // Note: Syntax is not final!
    let result = deer::json::from_value::<Body>(payload);
    let error = result.expect_err("should fail");
    println!("{error:?}");

    let result = serde_json::from_value::<Body>(payload);
    let error = result.expect_err("should fail");
    println!("{error:?}");
}
```

`serde` will fail immediately upon encountering that `256` is larger than what `u8` allows. This leads to frustration for the API consumer, as once they fix that issue the next problem, that `string` cannot be null, will be returned. `serde` also does not include path information about where the issue is located, `deer` does!

`deer` solves this problem, by returning every issue present. This means that a single API call with the payload given will result in the errors: `256 larger than u8::MAX`, `null is not String`, and `extra key "extra" provided`.

> This in turn also means that deer can be used for custom validation while deserializing, while still being able to return all validation issues.

Reviewer: ```suggestion
This in turn also means that `deer` can be used to implement custom validation while deserializing, while still being able to return all validation issues.
```

<sub>

> deer might provide a way in the future to describe these constraints.

Reviewer: Not sure how it'll behave inside an HTML element but should we have `deer` code-styled like the rest of the doc?

</sub>

## Limitations

> deer currently does **not** parse values itself, but relies on external parsers like `serde_json`, this means that parsing will be fail-fast, and deer only touches syntactically correct values.

Reviewer: ```suggestion
`deer` currently does **not** parse values itself, but relies on external parsers like `serde_json`, this means that parsing will be fail-fast, and `deer` only touches syntactically correct values.
```

<sup>
How the introspected format might look like is currently unknown.
</sup>

### Validation

Currently, to be able to do validation, one must create a new type that performs the validation step. This extra boilerplate is often a pain point. The idea is to instead allow for _optional_ validation via combinators using the derive macro.

<details>
<summary>What it may look like</summary>

```rust
#[derive(deer::Deserialize)]
struct Payload {
    #[validate(all(min(12), max(24)))]
    length: u8
}
```

</details>

### Lax Deserialization

Instead of strictly deserializing types, one might prefer to deserialize while coercing values (`"1"` might be interpreted as `1` instead). This behavior would be opt-in instead of opt-out and enabled on the `derive` level.

### Deserializer with Parser

`deer` currently relies on external tools (notably `serde`) to implement parsing of different formats, which means that we still fail-fast during the parsing of malformed input. In the future, `deer` might provide its own parser, which tries to recover from parsing errors and still provide parsing diagnostics.

<details>
<summary>What it may look like</summary>

```json
{
  "i8": "string"

> ```

Reviewer: Would us picking something other than JSON avoid this? 
<img width="788" alt="image" src="https://user-images.githubusercontent.com/25749103/194579717-8301d1dc-95cb-4d3d-b29c-facede10a27e.png">

Because I think we're going to have to deal with this on every PR for this text, not just this one.

Author: I removed formatting for now as I wasn't able to find a way to display incomplete JSON otherwise ([99538ec](https://github.com/hashintel/hash/pull/1163/commits/99538ec20ea56c29dc8876b1ebfc8fc61cf204c0))

Reviewer: Dang that's quite annoying. Perhaps we can revisit later on

`deer` currently relies on external tools (notably `serde`) to implement parsing of different formats, which means that we still fail fast during the parsing of malformed input. In the future, `deer` might provide a parser that tries to recover from parse errors and provide meaningful diagnostics.

</details>

### Level of "`deer`"

`deer` strives to be as composable and configurable as possible, which means that all non-core behavior should be opt-in, and the speed penalty of bare `deer` should be minimal. In the future, we might want to provide additional configuration parameters (maximum depth, the maximum number of errors, etc.) to allow for further tweaking in all use cases where speed is of utmost importance.

## Contributors

> `deer` was created by [Bilal Mahmoud](https://github.com/indietyp). It is being developed in conjunction with [HASH](https://hash.dev/). As an open-source project, we gratefully accept external contributions and have published a [contributing guide](https://github.com/hashintel/hash/blob/main/CONTRIBUTING.md) that outlines the process. If you have questions, please reach out to us on our [Discord server](https://hash.ai/discord).

Reviewer 2: ```suggestion
`deer` was created by [Bilal Mahmoud](https://github.com/indietyp). It is being developed in conjunction with [HASH](https://hash.dev/). As an open-source project, we gratefully accept external contributions and have published a [contributing guide](https://github.com/hashintel/hash/blob/main/CONTRIBUTING.md) that outlines the process. If you have questions, please reach out to us on our [Discord server](https://hash.ai/discord). You can also report bugs [directly on the GitHub repo](https://github.com/hashintel/hash/issues/new?assignees=Alfred-Mountfield%2CTimDiekmann%2Cindietyp&labels=A-deer%2CC-bug&template=bug-report-deer.yml).
```

## License

> Licensed under either of [Apache License, Version 2.0](LICENSE-APACHE.md)or [MIT license](LICENSE-MIT.md) at your option.

Reviewer 3: ```suggestion
Licensed under either of [Apache License, Version 2.0](LICENSE-APACHE.md) or [MIT license](LICENSE-MIT.md) at your option.
```

## Example

A popular example for `deer` are end-user facing APIs, given the following example:

```rust
#[derive(Debug, serde::Deserialize, deer::Deserialize)]
struct Body {
    u8: u8,
    string: String
}

fn main() {
    let payload = json!({
        "u8": 256,
        "string": null,
        "extra": 1
    });

    // Note: Syntax is not final!
    let result = deer::json::from_value::<Body>(payload);
    let error = result.expect_err("should fail");
    println!("{error:?}");

    let result = serde_json::from_value::<Body>(payload);
    let error = result.expect_err("should fail");
    println!("{error:?}");
}
```

`serde` will fail immediately upon encountering that `256` is larger than what `u8` allows. This leads to frustration for the API consumer, as once they fix that issue the next problem, that `string` cannot be null, will be returned. `serde` also does not include path information about where the issue is located, `deer` does!

> `deer` solves this problem, by returning every issue present.
> This means that a single API call with the payload given will result in the errors: `256 larger than u8::MAX`, `null is not String`, and `extra key "extra" provided`.

Reviewer 3: ```suggestion
`deer` solves this problem, by returning every issue present. This means that a single API call with the payload given will result in the errors: `256 larger than u8::MAX`, `null is not String`, and `extra key "extra" provided`.
```

This in turn also means that deer can be used for custom validation while deserializing, while still being able to return all validation issues.

<sub>
deer might provide a way in the future to describe these constraints.
</sub>

## Limitations

deer currently does **not** parse values itself, but relies on external parsers like `serde_json`, this means that parsing will be fail-fast, and deer only touches syntactically correct values.

## Future Plans

> The first release of `deer` is intentionally minimal and tries to lay a good foundation to extend functionality in the future.
> There are many future possible directions and ideas we're trying to see if they can benefit the different use cases.

Reviewer 3: ```suggestion
The first release of `deer` is intentionally minimal and tries to lay a good foundation to extend functionality in the future. There are many future possible directions and ideas we're trying to see if they can benefit the different use cases.
```

### Introspection Support

Currently, popular crates like `serde` do not provide a way to introspect what the output will be. Other tools try to fill the gap by manually interpreting the instructions given to these crates. This often leads to edge cases, resulting in the dissonance between the expected value and reality. The goal with the explicit support of introspection is to allow other tools to make use of it and build abstractions around it instead of trying to reverse engineer.

<sup>
How the introspected format might look like is currently unknown.
</sup>

### Validation

Currently, to be able to do validation, one must create a new type that performs the validation step. This extra boilerplate is often a pain point. The idea is to instead allow for _optional_ validation via combinators using the derive macro.

<details>
<summary>What it may look like</summary>

```rust
#[derive(deer::Deserialize)]
struct Payload {
    #[validate(all(min(12), max(24)))]
    length: u8
}
```

</details>

### Lax Deserialization

> Instead of strictly deserializing types, one might prefer to deserialize while coercing values (`"1"` might be interpreted as `1` instead),
> This behavior would be opt-in instead of opt-out and enabled on the `derive` level.

Reviewer 3: ```suggestion
Instead of strictly deserializing types, one might prefer to deserialize while coercing values (`"1"` might be interpreted as `1` instead). This behavior would be opt-in instead of opt-out and enabled on the `derive` level.
```

### Deserializer with Parser

> `deer` currently relies on external tools (notably `serde`) to implement parsing of different formats, which means that we still fail-fast during the parsing of malformed input.
> In the future, `deer` might provide its own parser, which tries to recover from parsing errors and still provide parsing diagnostics.

Reviewer 3: ```suggestion
`deer` currently relies on external tools (notably `serde`) to implement parsing of different formats, which means that we still fail-fast during the parsing of malformed input. In the future, `deer` might provide its own parser, which tries to recover from parsing errors and still provide parsing diagnostics.
```

<details>
<summary>What it may look like</summary>

```json
{
  "i8": "string"
```

> `deer` currently relies on external tools (notably `serde`) to implement parsing of different formats, which means that we still fail fast during the parsing of malformed input.
> In the future, `deer` might provide a parser that tries to recover from parse errors and provide meaningful diagnostics.

Reviewer 3: ```suggestion
`deer` currently relies on external tools (notably `serde`) to implement parsing of different formats, which means that we still fail fast during the parsing of malformed input. In the future, `deer` might provide a parser that tries to recover from parse errors and provide meaningful diagnostics.
```

</details>

### Level of "`deer`"

> `deer` strives to be as composable and configurable as possible, which means that all non-core behavior should be opt-in, and the speed penalty of bare `deer` should be minimal.
> In the future, we might want to provide additional configuration parameters (maximum depth, the maximum number of errors, etc.) to allow for further tweaking in all use cases where speed is of utmost importance.

Reviewer 3: ```suggestion
`deer` strives to be as composable and configurable as possible, which means that all non-core behavior should be opt-in, and the speed penalty of bare `deer` should be minimal. In the future, we might want to provide additional configuration parameters (maximum depth, the maximum number of errors, etc.) to allow for further tweaking in all use cases where speed is of utmost importance.
```

> # deer

Reviewer 3: In markdown, we use softwraps instead of hardbreaks, so we basically have one line per paragraph. Could you adjust this file please?

Author: fixed in [e0bfcd6](https://github.com/hashintel/hash/pull/1163/commits/e0bfcd65d688297910bb00fcb76911a517a4804c)

`deer` is a backend agnostic deserialization framework, featuring meaningful error
messages and context (utilizing [`error-stack`](https://crates.io/crates/error-stack)) and
a fail-slow behavior by default.

## Fail-Slow

Currently available Rust deserializers have mostly been developed with correctness and
speed in mind. These are universally beneficial optimizations, but in certain cases (such
as when collecting user-facing validation feedback) there are relatively few options
available within Rust that allow for extended evaluation beyond a single error. `deer`
aims to improve this situation by consciously trading off an acceptable degree of speed to
enable the surfacing of multiple errors.

## Example

A popular example for `deer` are end-user facing APIs, given the following example:

```rust

> #[derive(serde::Deserialize, deer::Deserialize)]

Reviewer 3: `expect_err` requires `Debug`
```suggestion
#[derive(Debug, serde::Deserialize, deer::Deserialize)]
```

struct Body {
    i8: i8,
    string: String
}

fn main() {
    let payload = json!({

>         "i8": 257,

Reviewer 3: Maybe, we want to use `u8` and `256`? That's an error, which happens more often by accident.

Author: fixed in [e0bfcd6](https://github.com/hashintel/hash/pull/1163/commits/e0bfcd65d688297910bb00fcb76911a517a4804c) and [5cec334](https://github.com/hashintel/hash/pull/1163/commits/5cec334dfd86f05fc9f46de8a6a770d89cf30827)

        "string": null,
        "extra": 1
    });

    // Note: Syntax is not final!

>     let result = deer::json::from_string::<Body>(payload);

Reviewer 3: To be consistent
```suggestion
    let result = deer::json::from_value::<Body>(payload);
```

>     let error = result.expect_error("should fail");

Reviewer 3: ```suggestion
    let error = result.expect_err("should fail");
```

    println!("{error:?}");

>     let result = serde_json::from_str::<Body>(payload);

Reviewer 3: ```suggestion
    let result = serde_json::from_value::<Body>(payload);
```

>     let error = result.expect_error("should fail");

Reviewer 3: ```suggestion
    let error = result.expect_err("should fail");
```

## Future Plans

The first release of `deer` is intentionally minimal and tries to lay a good foundation to
extend functionality in the future.
There are many future possible directions and ideas we're trying to see if they can
benefit the different use cases.

### Introspection Support

Currently, popular crates like `serde` do not provide a way to introspect what the output
will be. Other tools try to fill the gap by manually interpreting the instructions given
to these crates. This often leads to edge cases, resulting in the dissonance between the
expected value and reality. The goal with the explicit support of introspection is to
allow other tools to make use of it and build abstractions around it instead of trying to
reverse engineer.

<sup>
How the introspected format might look like is currently unknown.
</sup>

### Validation

Currently, to be able to do validation, one must create a new type that performs the
validation step. This extra boilerplate is often a pain point. The idea is to instead
allow for _optional_ validation via combinators using
the derive macro.

<details>
<summary>What it may look like</summary>

```rust
#[derive(deer::Deserialize)]
struct Payload {

>     #[validate(min(12) & max(24))]

Reviewer 3: nit-pick: Do we want to use `cfg` notation?
```suggestion
    #[validate(all(min(12), max(24)))]
```

    length: u8
}
```

</details>

### Lax Deserialization

Instead of strictly deserializing types, one might prefer to deserialize while coercing
values (`"1"` might be interpreted as `1` instead),
This behavior would be opt-in instead of opt-out and enabled on the `derive` level.

### Deserializer with Parser

`deer` currently relies on external tools (notably `serde`) to implement parsing of
different formats, which means that we still fail-fast during the parsing of malformed
input.
In the future, `deer` might provide its own parser, which tries to recover from parsing
errors and still provide parsing diagnostics.

<details>
<summary>What it may look like</summary>

```json
{
  "i8": "string"
```

`deer` currently relies on external tools (notably `serde`) to implement parsing of
different formats, which means that we still fail fast during the parsing of malformed
input.
In the future, `deer` might provide a parser that tries to recover from parse errors and

> provide meaningful diagnostics.

Reviewer 3: This section is almost a duplicate of the above

Author: kind of. I want to convey that the current limitation of failing fast on parsing might be lifted in the future. Maybe rewording is needed.

</details>

### Level of "`deer`"

`deer` strives to be as composable and configurable as possible, which means that all
non-core behavior should be opt-in, and the speed penalty of bare `deer` should be
minimal.
In the future, we might want to provide additional configuration parameters (maximum
depth, the maximum number of errors, etc.) to allow for further tweaking in all use cases
where speed is of utmost importance.

## Contributors

`deer` was created by [Bilal Mahmoud](https://github.com/indietyp). It is being developed
in conjunction with [HASH](https://hash.dev/). As an open-source project, we gratefully
accept external contributions and have published
a [contributing guide](https://github.com/hashintel/hash/blob/main/CONTRIBUTING.md) that
outlines the process. If you have questions, please reach out to us on

> our [Discord server](https://hash.ai/discord).

Reviewer 3: We also maybe want to link to issues and discussions. cc @nonparibus

# Deer

`deer` is a backend agnostic deserialization framework, featuring meaningful error
messages and context (utilizing [`error-stack`](https://crates.io/crates/error-stack)) and
a fail-slow behavior by default.

## Fail-Slow

Currently, Rust deserializers are good at one thing: parsing a lot of data correct and
*fast*. This is often what is desired, but the need for speed sacrifices usability in
scenarios where multiple errors are of interest.
`deer` solves this exact problem, by allowing multiple errors, trying as hard as possible
not to fail at the first error.

## Example

A popular example for `deer` are end-user facing APIs, given the following example:

```rust
#[derive(serde::Deserialize, deer::Deserialize)]
struct Body {
    i8: i8,
    string: String
}

fn main() {
    let payload = json!({
        "i8": 257,
        "string": null,
        "extra": 1
    });
}
```

> `serde` will fail immediately upon encountering that `257` is larger than what `i8`

Author: fixed in [3de3e3d](https://github.com/hashintel/hash/pull/1163/commits/3de3e3deecd396482d70a399309d847aedb31ef3)

# deer

`deer` is a backend agnostic deserialization framework, featuring meaningful error
messages and context (utilizing [`error-stack`](https://crates.io/crates/error-stack)) and
a fail-slow behavior by default.

## Fail-Slow

> The current landscape of Rust deserializers has been developed with correctness and speed
> in mind.
> This is often what is desired but means that certain use-cases like user-facing errors
> are inadequately supported in the Rust ecosystem.
> `deer` tries to solve this problem, by trading off an acceptable amount of speed to enable
> the surfacing of multiple errors.

Reviewer 2: ```suggestion
Currently available Rust deserializers have mostly been developed with correctness and speed in mind. These are universally beneficial optimizations, but in certain cases (such as when collecting user-facing validation feedback) there are relatively few options available within Rust that allow for extended evaluation beyond a single error. `deer` aims to improve this situation by consciously trading off an acceptable degree of speed to enable the surfacing of multiple errors.
```

Builds on https://github.com/hashintel/hash/pull/1163#discussion_r988455081

> # Deer

Reviewer 2: I don't _think_ we want to capitalize this here, as we don't anywhere else (similar to `error-stack`'s README](https://github.com/hashintel/hash/tree/dw/mixed-updates/packages/libs/error-stack#error-stack))

Author: fixed in [2bb0b1b](https://github.com/hashintel/hash/pull/1163/commits/2bb0b1b818abf9c9c3622249777410805a730abc)

`deer` is a backend agnostic deserialization framework, featuring meaningful error
messages and context (utilizing [`error-stack`](https://crates.io/crates/error-stack)) and
a fail-slow behavior by default.

## Fail-Slow

> Currently, Rust deserializers are good at one thing: parsing a lot of data correct and

Reviewer 2: It occurs to me that this might come across as unnecessarily harsh. One might believe that Rust deserializers are good at more than _just_ one thing.

We should still make the case that most Rust deserializers have been developed with correctness and speed in mind, but can avoid saying they are "good at one thing", as this runs the risk of unintentionally sounding exclusive/exhaustive.

Author: > It occurs to me that this might come across as unnecessarily harsh.

I guess that's where my inner German surfaced^^

I changed the wording in [2bb0b1b](https://github.com/hashintel/hash/pull/1163/commits/2bb0b1b818abf9c9c3622249777410805a730abc), is this better?

_fast_. This is often what is desired, but the need for speed sacrifices usability in
scenarios where multiple errors are of interest.

> `deer` solves this exact problem, by allowing multiple errors, trying as hard as possible

Reviewer 2: We might explicitly preface this by saying something to the effect of "`deer` approaches deserialization differently, trading off an acceptable amount of speed to enable the surfacing of multiple errors".

Author: changed in [2bb0b1b](https://github.com/hashintel/hash/pull/1163/commits/2bb0b1b818abf9c9c3622249777410805a730abc), is this better?

struct Body {
    i8: i8,
    string: String
}

fn main() {
    let payload = json!({
        "i8": 257,
        "string": null,
        "extra": 1
    });

    // Note: Syntax is not final!
    let result = deer::json::from_string::<Body>(payload);
    let error = result.expect_error("should fail");
    println!("{error:?}");

    let result = serde_json::from_str::<Body>(payload);
    let error = result.expect_error("should fail");
    println!("{error:?}");
}
```

`serde` will fail immediately upon encountering that `257` is larger than what `i8`
allows. This leads to frustration for the API consumer, as once they fix that issue the
next problem, that `string` cannot be null, will be returned. `serde` also does not include
path information about where the issue is located, `deer` does!

`deer` solves this problem, by returning every issue present.
This means that a single API call with the payload given will result in the
errors: `257 larger than i8::MAX`, `null is not String`, and `extra key "extra" provided`.

This in turn also means that deer can be used for custom validation while
deserializing, while still being able to return all validation issues.

<sub>

> deer might provide a way in the future to describe these constraints.

Reviewer 2: On the topic of possible futures, is there a world in which we might tweak the actual amount of speed users are willing to trade-off (i.e. have multiple strategies for allowing errors, some more performant vs exhaustive than others?) I can see wanting to tweak this from time to time, depending on implementation context, and it would be nice to provide 'degrees of deer' rather than just 'deer or no deer'.

Author: Yes, my goal is still to be as fast as possible (there will be overhead compared to serde, but I want to benchmark before optimizing), but I could imagine a future where you can opt into some checks while excluding others.

If deer goes forward with integrated validation or coercion (in 0.2?) I want to preserve the ability to altogether opt-out of that behavior if desired.

Reviewer 2: Do you want to add this in to the README? I would be inclined to flesh out 'future plans' as much as possible here, in the initial release, to get people excited/potentially attract likeminded others.

Author: I added a section on future plans in [644ba3b](https://github.com/hashintel/hash/pull/1163/commits/644ba3b2cca3651d8096559b65ee21df7f34afc3) and [871d1bd](https://github.com/hashintel/hash/pull/1163/commits/871d1bdb37f7d6cb8677002c28ee3f870eae6934).

</sub>

## Limitations

deer currently does **not** parse values itself, but relies on external parsers
like `serde_json`, this means that parsing will be fail-fast, and deer only touches
syntactically correct values.

## Contributors

`deer` was created by [Bilal Mahmoud](https://github.com/indietyp). It is being developed in conjunction with [HASH](https://hash.dev/). As an open-source project, we gratefully accept external contributions and have published a [contributing guide](https://github.com/hashintel/hash/blob/main/CONTRIBUTING.md) that outlines the process. If you have questions, please reach out to us on our [Discord server](https://hash.ai/discord).

> <!-- The license section is in line with other Rust tools, like serde -->
> <!-- markdownlint-disable MD001 -->
> 
> #### License
> 
> <sup>
> Licensed under either of <a href="LICENSE-APACHE">Apache License, Version
> 2.0</a> or <a href="LICENSE-MIT">MIT license</a> at your option.
> </sup>
> 
> <br>
> 
> <sub>
> Unless you explicitly state otherwise, any contribution intentionally submitted
> for inclusion in this crate by you, as defined in the Apache-2.0 license, shall
> be dual licensed as above, without any additional terms or conditions.
> </sub>

Reviewer 2: ```suggestion
## License

Licensed under either of [Apache License, Version 2.0](LICENSE-APACHE.md) or [MIT license](LICENSE-MIT.md) at your option.

For more information about contributing to this crate, see our top-level [CONTRIBUTING](https://github.com/hashintel/hash/blob/main/CONTRIBUTING.md) policy.
```

Sorry to reformat this away from serde-style to better match our other READMEs. The latter para has also been changed for pedantic accuracy (while the code contributed will be released upon acceptance as Apache 2.0/MIT, the act of contributing is still subject to our CLA).
